### ISSUE TITLE
Making module multi-language

### ISSUE THREAD

**Issue by juergenweb** [elabx/FieldtypeRecurringDates#6]
Hola Eduardo

First of all, I have not discovered any new issues - the module works as expected :-)

So I came up with a request: It would be great if you would make the module fully translatable.

At the moment there are only a view translatable strings included and most of the inputfield strings are not translatable. 
I have created a fork of your module at my GitHub account, where I implemented all the translations.

![Screenshot 2023-03-17 at 16-46-07 Bearbeite Seite Zuhause • webseite1 at](https://user-images.githubusercontent.com/9609925/225954506-25e17777-a5f1-4df7-8306-94ad511fa3ca.png)

As you can see all the texts are in German and they can be easily translated via the admin in the backend. :-)

Would be great if you include my changes inside your version, because I think other peoples will also be interested in.

The following files have been changed:

- InputfieldRecurringDates.js
- AlpineComponent.php
- InputfieldRecurringDates.module

You can find my fork [here](https://github.com/juergenweb/FieldtypeRecurringDates).

In AlpineComponent.php you have to add the textdomain next to the text. This is necessary to be able to load the translations, because this file will be integrated inside another file. Otherwise the translator will not find the translations.

The translations inside InputfieldRecurringDates.js will be defined inside the InputfieldRecurringDates.module. For this case I have created an new method, which holds the translation string called jsTranslations(). This is necessary to create Javascript variables via the PW $config->Js method.

I guess the best is you take a look at the files to see, what I have changed.

Have a nice weekend!

**Comment by elabx** [comment]
Woah nice! Thank you! I'll look into this, definitely want to make it translatable. If I recall correctly there is also a way to include translations in the module itself?

Do you think it might be worth to making the month and week names with IntlDateFormatter/strftime rather than hard coded translations?

Have a great weekend too and thanks for your support dedicated to the module! I greatly appreciate it!

**Comment by juergenweb** [comment]
Yes, there is a way to include the language files. I have included this now in my forked version too. You only have to create a folder called "languages" and add there the csv file for the language. I always name the file after the abbreviation of the language (de for German, fr for French and so on). 
You have to download each file of the module and copy them together in one file.  Take a look at how I did it.

Concerning the other question: I would leave it as hardcoded translateable strings - so the user has full flexibility and control over the text and it is always directly bind to the user language and not to the locale set.
I guess translating some months, weeks and days should not be a problem ;-)

**Comment by juergenweb** [comment]
I have now added the missing translations inside table of the created events:

![eeeee](https://user-images.githubusercontent.com/9609925/226126529-e62c220c-2774-44eb-95c5-e06f67c8e3f0.jpg)

Translations of number 1 are now in an extra file. I have called this file "*pw.php*" and it is located in the root of the module. This file is a copy of the "en.php" which is located inside the i18 folder of the RRule library. This file replaces all of the various translation files inside the i18 directory and make them all translateable via the PW admin. 
It contains over 80 translation strings, so it is very huge.
To be able to use this file I had made a change inside the AlpineComponent.php. Take a look [here](https://github.com/juergenweb/FieldtypeRecurringDates/blob/main/partials/AlpineComponent.php#L272). 
I pointed the location of the translation file to this custom file, so that all translations will be taken from this file.

The big advantage now is, that you are not longer limited by the languages supported by the library. I guess that there are 10 or 15 languages included, but if a user comes from a country which language is not supported, you will run into the first troubles.

The translations at number 2 (tableheaders) are also now translated inside the AlpineComponent.php.

Number 3 makes some troubles, because I cannot find where the text comes from. I know that the text will come from the JS but I do not know from which part. I mean especially the words "to" and "of". 
If the text ist directyl inside the Alpine.js, then I do not know if it could be translated or not. But I am sure you will know where the text phrases comes from. ;-)

I guess I have not forgotten anything to translate and the module will now be fully translated.
